helix-term/src/commands/lsp.rs

```diff
@@ -770,6 +721,131 @@ pub fn code_action(cx: &mut Context) {
     });
 }
 
+pub fn code_actions_for_range(
+    doc: &mut Document,
+    range: helix_core::Range,
+) -> Vec<(impl Future<Output = Result<Value, helix_lsp::Error>>, usize)> {
+    let mut seen_language_servers = HashSet::new();
+
+    doc.language_servers_with_feature(LanguageServerFeature::CodeAction)
+        .filter(|ls| seen_language_servers.insert(ls.id()))
+        // TODO this should probably already been filtered in something like "language_servers_with_feature"
+        .filter_map(|language_server| {
+            let offset_encoding = language_server.offset_encoding();
+            let lsp_range = range_to_lsp_range(doc.text(), range, offset_encoding);
+
+            language_server
+                .code_actions(
+                    doc.identifier(),
+                    lsp_range,
+                    // Filter and convert overlapping diagnostics
+                    lsp::CodeActionContext {
+                        diagnostics: doc
+                            .diagnostics()
+                            .iter()
+                            .filter(|&diag| {
+                                range.overlaps(&helix_core::Range::new(
+                                    diag.range.start,
+                                    diag.range.end,
+                                ))
+                            })
+                            .map(|diag| {
+                                diagnostic_to_lsp_diagnostic(doc.text(), diag, offset_encoding)
+                            })
+                            .collect(),
+                        only: None,
+                        trigger_kind: Some(CodeActionTriggerKind::INVOKED),
+                    },
+                )
+                .map(|request| (request, language_server.id()))
+        })
+        .collect::<Vec<_>>()
+}
+
+pub fn code_actions_on_save(
+    doc: &mut Document,
+) -> Option<
+    FuturesUnordered<impl Future<Output = Result<Vec<CodeActionOrCommandItem>, anyhow::Error>>>,
+> {
+    let code_actions_on_save_cfg = doc
+        .language_config()
+        .map(|c| c.code_actions_on_save.clone())?;
+
+    if code_actions_on_save_cfg.is_empty() {
+        return None;
+    }
+
+    let full_range = Range::new(0, doc.text().len_chars());
+
+    Some(
+        code_actions_for_range(doc, full_range)
+            .into_iter()
+            .map(|(request, ls_id)| {
+                let code_actions_on_save = code_actions_on_save_cfg.clone();
```

Comment:
This assumes that we can trigger multiple code actions in parallel, which works in some cases, but one action might impact another action (e.g. if A removes a function, and B formats the file). Maybe to simplify we should limit to a single code action on save (since that's the majority usecase anyway)

Reply:
If I'm understanding the LSP spec as long as the requests are in the correct order, the lsp should determine how best to execute those requests.

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#messageOrdering

Reply:
Right, but if you send actions A and B in parallel, the server is computing the edits based on that initial state. Server only considers the change applied when we send back a `didChange` event. Helix does use OT internally to transform edits (so initial -> A -> B can be applied) but it's not necessarily going to produce the same result as if we requested A, applied, then requested B and applied B.

Reply:
But over multiple language servers we can't guarantee the order. That's what you were getting at, I assume. Simplifying to a single code action would definitely make this problem go away, though then we are not feature parity with VSCode, if that is important.

Reply:
> Right, but if you send actions A and B in parallel, the server is computing the edits based on that initial state. Server only considers the change applied when we send back a `didChange` event. Helix does use OT internally to transform edits (so initial state -> A -> B can be applied) but it's not necessarily going to produce the same result as if we requested A, applied, then requested B and applied B.

Ok, I see. Thanks! Let me know which direction we want to take and I can make the changes.